qiskit/synthesis/clifford/clifford_decompose_layers.py

```diff
@@ -0,0 +1,392 @@
+)
+from qiskit.quantum_info.operators.symplectic.clifford_circuits import (
+    _append_h,
+    _append_s,
+    _append_cz,
+)
+
+
+def _default_cx_synth_func(mat, validate):
+    """
+    Construct the layer of CX gates from a boolean invertible matrix mat.
+    """
+    if validate:
+        if not check_invertible_binary_matrix(mat):
+            raise QiskitError("The matrix for CX circuit is not invertible.")
+
+    CX_circ = synth_cnot_count_full_pmh(mat)
+    CX_circ.name = "CX"
+
+    if validate:
+        _check_gates(CX_circ, ("cx", "swap"))
+
+    return CX_circ
+
+
+def _default_cz_synth_func(symmetric_mat, validate):
+    """
+    Construct the layer of CZ gates from a symmetric matrix.
+    """
+    nq = symmetric_mat.shape[0]
+    qc = QuantumCircuit(nq, name="CZ")
+
+    for j in range(nq):
+        for i in range(0, j):
+            if symmetric_mat[i][j]:
+                qc.cz(i, j)
+
+    if validate:
+        _check_gates(qc, "cz")
+    return qc
+
+
+def synth_clifford_layers(
+    cliff,
+    cx_synth_func=_default_cx_synth_func,
+    cz_synth_func=_default_cz_synth_func,
+    cx_cz_synth_func=None,
+    validate=False,
+):
+    """Synthesis of a Clifford into layers, it provides a similar decomposition to the synthesis
+    described in Lemma 8 of [1].
+
+    For example, a 5-qubit Clifford circuit is decomposed into the following layers:
+
+    .. parsed-literal::
+             ┌─────┐┌─────┐┌────────┐┌─────┐┌─────┐┌─────┐┌─────┐┌────────┐
+        q_0: ┤0    ├┤0    ├┤0       ├┤0    ├┤0    ├┤0    ├┤0    ├┤0       ├
+             │     ││     ││        ││     ││     ││     ││     ││        │
+        q_1: ┤1    ├┤1    ├┤1       ├┤1    ├┤1    ├┤1    ├┤1    ├┤1       ├
+             │     ││     ││        ││     ││     ││     ││     ││        │
+        q_2: ┤2 S2 ├┤2 CZ ├┤2 CX_dg ├┤2 H2 ├┤2 S1 ├┤2 CZ ├┤2 H1 ├┤2 Pauli ├
+             │     ││     ││        ││     ││     ││     ││     ││        │
+        q_3: ┤3    ├┤3    ├┤3       ├┤3    ├┤3    ├┤3    ├┤3    ├┤3       ├
+             │     ││     ││        ││     ││     ││     ││     ││        │
+        q_4: ┤4    ├┤4    ├┤4       ├┤4    ├┤4    ├┤4    ├┤4    ├┤4       ├
+             └─────┘└─────┘└────────┘└─────┘└─────┘└─────┘└─────┘└────────┘
+
+    Args:
+        cliff (Clifford): a clifford operator.
+        cx_synth_func (Callable): a function to decompose the CX sub-circuit.
+        cz_synth_func (Callable): a function to decompose the CZ sub-circuit.
+        cx_cz_synth_func (Callable): optional, a function to decompose both sub-circuits CZ and CX.
+        validate (Boolean): if True, validates the synthesis process.
+
+    Return:
+        QuantumCircuit: a circuit implementation of the Clifford.
+
+    Reference:
+        1. S. Bravyi, D. Maslov, *Hadamard-free circuits expose the
+           structure of the Clifford group*,
+           `arXiv:2003.09412 [quant-ph] <https://arxiv.org/abs/2003.09412>`_
+    """
+
+    cliff_cpy = cliff.copy()
```

Comment:
Is the copy here really needed? `cliff_cpy` is only passed to `_create_graph_state` and then it's being copied there as well. So one of those copies seem redundant